improv_ble_client/improv_ble_client.py

```diff
@@ -0,0 +1,469 @@
+        return str(self._ble_device.address)
+
+    @property
+    def name(self) -> str:
+        """Get the name of the device."""
+        return str(self._ble_device.name or self._ble_device.address)
+
+    @property
+    def rssi(self) -> int | None:
+        """Get the rssi of the device."""
+        if self._advertisement_data:
+            return self._advertisement_data.rssi
+        return None
+
+    async def can_identify(self) -> bool:
+        """Return if the device supports identify."""
+        _LOGGER.debug("%s: can_identify", self.name)
+
+        async def _can_identify() -> bool:
+            capabilities = await self.read_characteristic(
+                CHARACTERISTIC_UUID_CAPABILITIES
+            )
+            return bool(capabilities & prot.Capabilities.IDENTIFY)
+
+        return await self._execute(_can_identify)
+
+    async def identify(self) -> None:
+        """Identify the device."""
+        _LOGGER.debug("%s: identify", self.name)
+
+        async def _identify() -> None:
+            await self.send_cmd(prot.IdentifyCmd())
+
+        await self._execute(_identify)
+
+    async def need_authorization(self) -> bool:
+        """Return if the device needs authorization."""
+        _LOGGER.debug("%s: need_authorization", self.name)
+
+        async def _need_authorization() -> bool:
+            state = await self.read_characteristic(CHARACTERISTIC_UUID_STATE)
+            return state == prot.State.AUTHORIZATION_REQUIRED
+
+        return await self._execute(_need_authorization)
+
+    async def provision(
+        self,
+        ssid: str,
+        password: str,
+        state_callback: Callable[[prot.State], None] | None,
+    ) -> str | None:
+        """Provision the device.
+
+        Returns the redirect url or None.
+        """
+        _LOGGER.debug("%s: provision ssid: %s, pw: %s", self.name, ssid, password)
+
+        async def _provision() -> str | None:
+            """Execute the procedure"""
+
+            def handle_error(value: prot.Error) -> None:
+                error_fut.set_result(value)
+
+            def handle_state(state: prot.State) -> None:
+                if state_callback:
+                    state_callback(state)
+
+            subscriptions = [
+                self._notification_handlers.subscribe_error(handle_error),
+                self._notification_handlers.subscribe_state(handle_state),
+            ]
+            error_fut: asyncio.Future[prot.Error] = asyncio.Future()
+            provisioned_fut = self.receive_response(prot.WiFiSettingsRes)
+
+            try:
+                await self.send_cmd(
+                    prot.WiFiSettingsCmd(bytes(ssid, "utf-8"), bytes(password, "utf-8"))
+                )
+
+                done, pending = await asyncio.wait(
+                    (error_fut, provisioned_fut),
+                    return_when=asyncio.FIRST_COMPLETED,
+                )
+                for task in pending:
+                    task.cancel()
```

Comment:
This might complain about an unretrieved exception if it's not awaited

Reply:
The variable name is misleading, these are futures, not tasks.
I think the potential error here is if the cancelled futures have their result set after we cancel them?

Reply:
> I think the potential error here is if the cancelled futures have their result set after we cancel them?

That is a risk if you await them as they will propagate cancelation from the point of await

Reply:
As discussed on Discord the code is fine, but the variable names are misleading and unclear. I'll clean it up.